### PR TITLE
Initial commit of KV implementation.

### PR DESCRIPTION
**Pull request by reterVision**

### CODE REVIEW

File: kv/src/main/java/com/test/kv/interfaces/Storage.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Interface for storage
+ */
+public interface Storage<KeyType, ValueType> {
+    /**
+     * Get method returns the value of user specified key in storage if it exists
+     *
+     * @throws KVKeyDoesNotExistsException when key is not presented in storage
+     */
+    ValueType Get(KeyType key) throws KVKeyDoesNotExistsException;
```
Comment (jason-xxl) [inline kv/src/main/java/com/test/kv/interfaces/Storage.java]: since it throw exception on key missing, I think the interface need another .HasKey method to allow test key existence.

File: kv/src/main/java/com/test/kv/lib/storage/Memory.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+    /**
+     * maxKeySize is the max number of keys storage can store
+     */
+    protected int maxKeySize = 1024;
+
+    /**
+     * expireAtStorage stores the key->TTL pair for cleanup expired keys usage
+     */
+    protected Map<KeyType, Long> expireAtStorage;
```
Comment (jason-xxl) [inline kv/src/main/java/com/test/kv/lib/storage/Memory.java]: I think the data structure that group cache using is simpler.
https://github.com/golang/groupcache/tree/master/lru
since it's LRU already, I have a feeling that TTL is bit redundant. just feeling anyway.

ignore me if this is for storage not cache.

File: kv/src/main/java/com/test/kv/lib/storage/Memory.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+import com.sun.org.apache.bcel.internal.generic.INSTANCEOF;
+import com.sun.tools.jdi.ConcreteMethodImpl;
+import com.test.kv.exceptions.KVKeyDoesNotExistsException;
+import com.test.kv.interfaces.Storage;
+
+/**
+ * Implementation of Storage -- In memory store with LRU
+ */
+public class Memory<KeyType, ValueType> implements Storage<KeyType, ValueType> {
```
Comment (jason-xxl) [inline kv/src/main/java/com/test/kv/lib/storage/Memory.java]: Memory is bit too general as a type name?

File: kv/src/main/java/com/test/kv/lib/storage/Memory.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            }
+        }
+    }
+
+    /**
+     * storeKeyAndTTL stores the key with its ttl in expireAtStorage
+     */
+    private void storeKeyAndTTL(KeyType key, Long ttlInSec) {
+        Long expiredAt = new Long(Instant.now().getEpochSecond() + ttlInSec.longValue());
+        expireAtStorage.put(key, expiredAt);
```
Comment (jason-xxl) [inline kv/src/main/java/com/test/kv/lib/storage/Memory.java]: (trying to be very strict here, may not necessary in practice..)
since store need to touch 3 different data object without transaction protection, it got potential to result in ABA problem when frequently set and remove same key.

(that's also one reason that group cache's solution looks simpler.)

Comment (jason-xxl) [inline kv/src/main/java/com/test/kv/lib/storage/Memory.java]: It may need some memory consumption control if not sufficient memory?

File: kv/src/main/java/com/test/kv/lib/storage/Memory.java (hunk 3 above)
Comment (reterVision) [inline kv/src/main/java/com/test/kv/lib/storage/Memory.java]: yeah, looks like so, I'm getting too used to `package.Type` calling pattern in Go :trollface:

File: kv/src/main/java/com/test/kv/lib/storage/Memory.java (hunk 4 above)
Comment (reterVision) [inline kv/src/main/java/com/test/kv/lib/storage/Memory.java]: True, I'm actually using this simple project to get myself familiar with possible edge cases so I don't handle edge cases very well... will improve!

**Comment by reterVision** [comment]
Thanks for the review! @jason-xxl
